Pybind11 integration

TanDro opened the pull request:
* Add Pybind11 functionality to pymoskito
* New example 'Tanksystem

TanDro committed:
Improve plot feature to show color labels

TanDro committed:
Improve png export

TanDro committed:
Add example for cpp implementation, add pybind11 as lib, start implementation of cpp controller class

TanDro committed:
Add example for cpp implementation, add pybind11 as lib, start implementation of cpp controller class

TanDro committed:
Add export of datapoints directly from listwidget

TanDro committed:
Update example for cpp pid implementation

TanDro committed:
First implementation of Cpp module for controller with running of make

TanDro committed:
Update test scenario to 2 tanks, add first approach for observer, Add write and test of CMakeLists.txt

TanDro committed:
Update cpp integration to run under windows

TanDro committed:
Improve plot coloring

TanDro committed:
Add high gain observer to tanksystem example, update code a bit

TanDro committed:
Make Observer module compileable, can't return array

TanDro committed:
Add c++ version of observer for pymoskito with wrapper functions

cklb committed:
Added try/except clause on module init

cklb committed:
Merge branch 'master' into develop

TanDro committed:
Add exception handling at init of modules before simulation start

TanDro committed:
Add colors to log dock

TanDro committed:
Add colors to log dock

TanDro committed:
First possible solution for install with pybind11

cklb committed:
Moved log color definitions into config file

cklb committed:
Changed logging massages in pendulum example

TanDro committed:
Remove inheritance classes for cpp from SimualtionModule, Format controller and observer code, Remove copyright

TanDro committed:
Move generateBinding to tools, als helper function,
At init only the make process is started for the c code,
Remove binding keyword (not necessary)

cklb committed:
Some thoughts on the bindings:
* Refactored the src files to contain one class per file
* Renamed the Controller class to fit the name of its binding
  (Observer is still to be done)
* Moved the binding generation out of the registration process.
  The registry module should not have anything to do with something else
  than registration.
* The introduced extra property `Module` not quite meaningful and should
  not be the way to do things. The whole process should not depend on
  whether a pure python or a wrapper module is used.
* Refactored the binding generation in the `tools` module to get it pep8
  conform and to remove code duplicates.
* As it seems to me atm an installation of `pybind11` via pip misses the
  required `.cmake` files for the `find_package(pybind11)` call to
  succeed. I'm opening an issue upstream to get further information on
  this topic.
* After a little though on the topic, I really dislike the variant on
  using pybind as a git submodule since it greatly complicates the
  installation process of pymoskito for an optional dependency.
* Added a pybind entry to the requirements

TanDro committed:
Fix tooltip of new export button

TanDro commented:
Based on your commit message of adc6ec74c4a16e1639b8a74291bc9d4c373e61ce:

* If it would be possible to handle the pybind install as a dependency, this would be nice. Where do you create the issue?
* I'm not sure, how it can realized without a wrapper module. It is needed to transport some informations from the simulation circle to a c class. I see two possible options: 

1. A python module: Makes it easy for the existing process. The user can do the wrapping of settings and in- and output data handling for his external class.
2. A c++ template class with an internal python wrapper: Better encapsulation of the functionality, but I mean, it is not easy to implement. The handling of in- and output data and the settings must done in it.

What am I missing? Maybe there is a better solution...

cklb commented:
After the recommended first approach of asking in the upstream gitter chat yielded no answer, I opened an issue here: https://github.com/pybind/pybind11/issues/1628 .
Regarding your second point: I like the idea of a python wrapper, the point is the all the compilation and configuration stuff should be done outside the simulator or in other words: No changes the the `simulation_core`, `simulation_interface`, `simulation_gui` or `registry` modules should be made.

cklb commented on pymoskito/tools.py:
```diff
@@ -264,3 +268,148 @@ def get_figure_size(scale):
+        f.write(c_make_lists)
+
+
+class CSVExporter(object):
```
This should also be removed form this PR.

cklb commented on pymoskito/tools.py:
```diff
@@ -264,3 +268,148 @@ def get_figure_size(scale):
+    add_binding_config(c_make_lists_path, module_name)
+
+    # build
+    if os.name == 'nt':
```
This will start a build process every time the wrapping class is loaded, which is done when the application is started.
Is this sufficient or would you like to retain the possibility to issue a new build when the object is actually instantiated?

cklb commented on pymoskito/simulation_modules.py:
```diff
@@ -67,6 +70,24 @@ def __init__(self, settings):
         self._settings["step width"] = None
         self._settings.pop("modules", None)
 
+        if settings.get("Module", None) is not None:
```
I think, if the YIT compilation is needed, the `generate_binding` should be split into `create_binding_config` and `build_binding` . Then a mix-in class like `CppBinding` can be added which adds this functionality and is provided with the required information. Afterwards you can  wirte something along the lines of `class PIDController(pm.Controller,  pm.CppBinding):` in your code.
If you do not dislike this approach to much, I'll provide an implementation of that idea.

TanDro replied:
Yes I think the YIT sould be avaiable and I like your idea with the CppBinding Class. First approach is seen in new commit.

cklb committed:
Some corrections to the last commit

* Fixed issues due to refactoring of .cpp files (Observer part is till missing)
* Provided better parameter name in `create_cmake_lists`

Using a system level pybind11 install, the `find-package` already works
very well.

TanDro committed:
Rename default regime name, Move binding functionality to own module, Derive PID class from controller and new binding class (same observer)

TanDro committed:
Fix binding typo

TanDro commented:
But now I have the problem when simulation finished
`10/12/2018 17:23:22 - SimulatorInteractor - INFO - Simulation reached 100%`
`Traceback (most recent call last):`
`  File "/home/jens/git/pymoskito/pymoskito/simulation_interface.py", line 529, in simulation_state_changed`
`    self._sim_data = copy.deepcopy(state_change.data)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 150, in deepcopy`
`    y = copier(x, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 240, in _deepcopy_dict`
`    y[deepcopy(key, memo)] = deepcopy(value, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 150, in deepcopy`
`    y = copier(x, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 240, in _deepcopy_dict`
`    y[deepcopy(key, memo)] = deepcopy(value, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 180, in deepcopy`
`    y = _reconstruct(x, memo, *rv)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 306, in _reconstruct`
`    value = deepcopy(value, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 150, in deepcopy`
`    y = copier(x, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 240, in _deepcopy_dict`
`    y[deepcopy(key, memo)] = deepcopy(value, memo)`
`  File "/home/jens/venvs/pymoskito/lib/python3.7/copy.py", line 169, in deepcopy`
`    rv = reductor(4)`
`TypeError: can't pickle TwoTankSystem objects`

TanDro committed:
Fix issue with deepcopy

TanDro committed:
Add method to instantiate needed class from binding

cklb commented on pymoskito/binding_modules.py:
```diff
@@ -0,0 +1,149 @@
+        with open(self.cmake_lists_path, "w") as f:
+            f.write(c_make_lists)
+
+    def get_module_instance(self, className):
```
Is it planned to actually use the constructor of the bindings (which is empty for now)  instead of calling the `create` method afterwards?
If this is not the case (which would be quite sad) the call to `get_module_instance` should actually return an instance of the object instead of its class.
In my opinion, the two calls could be combined into `init_binding` which would take all arguments for the constructor, instantiate the binding object and then store it in `self.binding` or something like that.

TanDro replied:
Yes you are right, the current situation is only for the initial used example. The create method should not used in general. There is not problem to return instance instead of the object. I will rewrite the example for this.

cklb committed:
Continued refactoring of cpp bindings

At the moment it seems like the import of the observer binding fails,
but this could also be caused by a dirty build directory on my side.

cklb committed:
Fixed remaining issue from refactoring

cklb committed:
Added export parameters to application config

cklb committed:
Refactored plot coloring

cklb committed:
Merge remote-tracking branch 'origin/colored_log_dock' into wip_improve_visu

# Conflicts:
#	pymoskito/simulation_gui.py
#	pymoskito/tools.py

cklb committed:
Migrated new settings

TanDro committed:
Remove pg exporter and use CSVExporter class also for export from plot

TanDro committed:
Update binding to return class, so a constructor can use, Update documentation

TanDro committed:
Add high res option for pyqt

TanDro committed:
Refactor exporter completly

TanDro committed:
Runnable version of new exporter with pandas data frame basis

TanDro committed:
Home path handling updated for windows

cklb committed:
Some small changes to the export function

These include:
* last export path and format is remembered
* units (if provided) are exported
* some pep8 related name changes

cklb committed:
Removed Qt dpi environment flag and added it to examples and docs

cklb committed:
Code cleanup for new features and new icons

cklb commented:
Today, I can happily announce that the pybind11 team managed to get the required headers to install via pip (https://github.com/pybind/pybind11/pull/2433).
Therefore, as soon as pybind11 v.2.6.0 is out, this PR can finally be merged into master.

TanDro commented:
That is good news.

TanDro committed:
Remove pybind as submodule

TanDro committed:
Update pybind version

TanDro committed:
Remove libs folder

TanDro committed:
Move binding to new version, Move to path library, include windows handling

TanDro committed:
Update tank example for current useage of cpp class binding

TanDro committed:
Rework example to run

cklb commented on MANIFEST.in:
```diff
@@ -8,6 +8,7 @@ include README.rst
 recursive-include pymoskito/resources *.py *.png *.yaml *.txt
 recursive-include pymoskito/examples *.py *.sreg
 recursive-include pymoskito/tests *.py
+recursive-include pymoskito/libs *
```
With the submodule gone, this and other statements can be deleted /reverted

cklb commented on pymoskito/binding_modules.py:
```diff
@@ -0,0 +1,176 @@
+        if module_path is None:
+            raise BindingException("Instantiation of binding class without"
+                                   " module_path is not allowed!")
+        self.module_path = Path(module_path)
```
The usage of pathlib really makes life easier

cklb commented on pymoskito/examples/tanksystem/controller.py:
```diff
@@ -0,0 +1,75 @@
+        ("Td", 12),
+        ("dt", 0.1),
+        ("output_limits", [0, 255]),
+        ("input_state", [0]),
```
This property is named after the built-in PID controller which in turn has a ambiguous name.
As the cpp implementation only supports scalar input anyway, this should be renamed or moved to __init__

cklb commented on pymoskito/examples/tanksystem/default.sreg:
```diff
@@ -0,0 +1,35 @@
+ # default regime file that is loaded when the simulation environment starts up
```
The default settings should yield an example which will not completely mess up the visualizer output

cklb commented on pymoskito/examples/tanksystem/visualization.py:
```diff
@@ -0,0 +1,310 @@
+try:
+    import vtk
+
+    class VtkTankVisualizer(pm.VtkVisualizer):
```
Either build a VTK variant or just delete it

cklb commented on pymoskito/examples/tanksystem/visualization.py:
```diff
@@ -0,0 +1,310 @@
+        self.axes.text(4 * st.rT1 + 3 * st.rT2, st.hT2 + 0.01, 'Tank 2', horizontalalignment='center')
+
+    def update_scene(self, x):
+        if x[0] > 0:
```
Handling of overflows? (non c- related ones)

cklb commented on requirements.txt:
```diff
@@ -9,3 +9,4 @@ PyQt5>=5.7.1
 control>=0.7.0
 Sphinx>=1.4.9
 sphinx-rtd-theme>=0.1.9
+pybind11>=2.6.0b1
```
Remember to bump once the release is out

TanDro replied:
yes, it is only for testing now and not the final version.

cklb commented on setup.py:
```diff
@@ -1,12 +1,21 @@
-    from distutils.core import setup
+    from distutils.core import setup, find_packages
+
+import subprocess, os
```
revert to original version

cklb reviewed (changes requested):
Just collected some remarks

cklb committed:
Updated python version to run tests for

cklb commented:
Finally, we should use the new example to add a short tutorial on how to include external c(++) code.

cklb committed:
Updated xfvb config

TanDro committed:
Remove submodule artefacts

TanDro committed:
Add boundary checks

TanDro committed:
Reduce count of fucntion, also unused methods, in c implementation for a better survey, Update documentation

TanDro committed:
Start with extra docu section for tanksystem

TanDro commented:
> Finally, we should use the new example to add a short tutorial on how to include external c(++) code.

I will revise the tank example, but I need some time for that.

cklb commented:
It seems like the pybind release will take some time, too. So no worries.
I'll draft an "external code" tutorial to begin with.

cklb committed:
Add required dependencies for xvfb

cklb committed:
Updated python version for packaging

cklb committed:
Migrate from testing via setup.py to unittest module

cklb committed:
Fix qt highdpi scaling warning

cklb committed:
Fix deprecated yaml loader

cklb committed:
Fix syntax warnings

cklb committed:
Added tests for Setpoint and Smooth Trajectories

Extended smooth transition to vectorial case

cklb committed:
Migrated the BallBeam example to new trajectories

cklb committed:
Migrated the BallTube example to new trajectories

cklb committed:
Migrated the Pendulum example to new trajectories

cklb committed:
Migrated the Simple-Pendulum example to new trajectories

cklb committed:
Fixed type error due to switching between different controllers in swing-up

cklb committed:
Adapted step-size in pendulum for full swing-up

cklb committed:
Introduced optional dependencies for vtk and pybind

TanDro committed:
Update parameter overview and first equations for tank docu

TanDro committed:
Add figure for tank

TanDro committed:
Update docu for tanks with equations and control and observer parts

TanDro committed:
Final equations added to docu for tank system, docu final

TanDro committed:
Update sim model, visu and observer

cklb commented:
Looks nice. Could you please rebase onto develop to get travis running again, so we can test cmake and friends in CI?

TanDro commented:
* pybind11 releases 2.6.0 https://github.com/pybind/pybind11/commit/59a2ac2745d8a57ac94c6accced73620d59fb844
* rebase done, but haven't check

cklb commented:
I wanted to add the following call to add the example to the test suite but it looks like I lost write access somehow.

```
--- a/pymoskito/tests/test_examples.py
+++ b/pymoskito/tests/test_examples.py
@@ -23,6 +23,9 @@ class TestExamples(unittest.TestCase):
     def test_car(self):
         self._test_example("car", debug=False)
 
+    def test_tanksystem(self):
+        self._test_example("tanksystem", debug=False)
+
     def _test_example(self, name: str, debug=False):
         """
         Run the given example while listening for errors on stdout
```

TanDro committed:
Move noise to sensor output

TanDro committed:
Fix equations of observer add boundary check

TanDro committed:
Update borders of controller

TanDro committed:
Add boundary checks for states

TanDro committed:
Update controller implementation

TanDro committed:
Add tanksystem to TestExamples of TestSuite

TanDro commented:
The test failed, but I can't reproduce the error on my test machines. The path is in either cases correct.
Edit: With a clean install, the error occurs.

TanDro committed:
Remove gitsubmodules file

cklb committed:
Fixed runtime warnings

cklb committed:
Restructured dependencies and updated travis config

cklb committed:
Improved error message and fixed path issue

cklb commented:
As I'm still not allowed to push anything, please see my changes in d388bdbc13bdeae313043e7b5843b3c25d6d87f5.

However,  after starting pymoskito and running the PID test regime, I get a SIGSEGV on closing the application. It seems like some memory is not beeing handled correctly.

cklb committed:
Fixed type in dep names

cklb commented:
Alright, I get it now. Will `binding_modules` look for the binding folder in '/home/travis/build/cklb/pymoskito/pymoskito/examples/tanksystem' as with my patch from above. However, some time back you did move the bindings folder into the utils folder of that project.
Thus, the folder cannot be found.

I'm not quite happy with the current project structure as this example shall act as a template for user implementations and should therefore be as simple as possible.
We could for example demand the user to put everything into the binding folder (.cpp, .h and .py files) or setup a `source` folder for the cpp files and leave teh bindings in the binding folder.

As the bindings and the c-code are tightly related, I would opt for one common folder.

TanDro committed:
Remove cmake building scripts from tools file

TanDro committed:
Add additionally logger messages

TanDro committed:
Add module path handling

TanDro committed:
Fix typo

TanDro committed:
Rename folder for better understanding

TanDro commented:
The main problem by one folder for the c-sources and the bindings that currently all generated files are also in this folder. Hence my intension was to have an binding folder with the bindings and the generated files. 
Maybe it could be a better solution to have all sources, bindings and the libs in one folder and compiling is in a seperate folder, I'm with you. I will make an offer in the next week.
Edit: Okay I did it now. @cklb Is this more the way you want?

TanDro committed:
Move binding files and remove binding folder

TanDro committed:
Restructure build process, tested with linux/mac, not tested with windows

cklb commented:
I like this way with separated build dir very much.

I added a few changes in the other branch:
* platform independent paths for the constructors
* refactored build dir name into a constant

The only "magical" thing left to know is that the corresponding binding files have to be `"binding_" + ${MOD_NAME}.cpp`. If we would require the user just to point to the binding file, the whole thing would get much less complicated.

cklb commented:
Currently test are failing because the sources do not get installed (I have to update `MANIFEST.in` to include them as an example).
However, as pymoskito will likely be installed in some path not writeable to the common user, we have to build the shared libraries on install and copy those over (only if the install target `[CPP]` is given). I will focus on that later.

TanDro committed:
Correct visu of heights in tanks

TanDro committed:
Add binding_class_name to reduce magic

TanDro committed:
Add h and cpp files of tests to manifest

simbach commented:
Running the **OpenLoop** regime on Windows leads to the following error: The _build folder is not generated
After a second run the _build folder is generated and the error message is: Could not load chache
Same behavior with **test_PID**

`28/10/2020 06:45:10 - SimulationGui - INFO - Simulating: OpenLoop
28/10/2020 06:45:10 - CppHighGainObserver - WARNING - CMakeLists.txt not found in module path.
28/10/2020 06:45:10 - CppHighGainObserver - INFO - Generating new CMake config.
28/10/2020 06:45:10 - CppHighGainObserver - INFO - Appending build info for 'HighGainObserver'
**CMake Error: The source directory "C:/Users/bachlers/sib/_GIT/pymoskito/pymoskito/examples/tanksystem/src/_build" does not exist**.
Specify --help for usage, or press the help button on the CMake GUI.
28/10/2020 06:45:10 - CppHighGainObserver - ERROR - Generation of binding config failed.
28/10/2020 06:45:10 - SimulatorInteractor - ERROR - Init of module Observer failed with
 "Generation of binding config failed."!

28/10/2020 06:45:10 - SimulatorInteractor - ERROR - Simulation Setup failed. Check Configuration.
28/10/2020 06:45:10 - SimulationGui - INFO - Simulation aborted
28/10/2020 06:45:10 - SimulationGui - INFO - Stopping Playback
28/10/2020 06:45:14 - SimulationGui - INFO - Simulating: OpenLoop
**Error: could not load cache**
28/10/2020 06:45:14 - CppHighGainObserver - ERROR - Build failed!
28/10/2020 06:45:14 - SimulatorInteractor - ERROR - Init of module Observer failed with
 "Build failed!"!
28/10/2020 06:45:14 - SimulatorInteractor - ERROR - Simulation Setup failed. Check Configuration.
28/10/2020 06:45:14 - SimulationGui - INFO - Simulation aborted
28/10/2020 06:45:14 - SimulationGui - INFO - Stopping Playback`

cklb committed:
Refactoring of binding module

cklb committed:
Refactoring of binding module

cklb committed:
Added first draft for CppBase docstring

cklb committed:
Added first draft for CppBase docstring

cklb commented:
I refactored the binding module to hopefully take care of all build folder issues (in the other repo).

Currently, there are 4 things open:
* @simbach could you provide a little insight into your setup (to explain the steps i the user manual) and test again with the new version? (I got rid of the `shell=True` flag which takes the shell interpreter out of the equation but is not yet tested under windows)
* @TanDro do you also experience a segfault after running a simulation with at least one cpp-module? 
* I have to tweak the install script to generate the bindings so that a user may test them when running the preinstalled examples.
* The current install call just copies the libs back into the original folder (I commented that out for the moment to easier clean the folder). How about the creation of a `lib` folder in the project root (same level as `src`) the keep the folder clean?

I added debug info in the build target to make finding errors easier for users. Once the culprit is found, I will introduce a flag for that.

simbach commented:
Some explanation to my setup:
* use winpython 3.7
* cmake version 3.14.19060802-MSVC_2 comming with visual studio 2019
* checked out the branch wip_pybind of the repo
* install the requirements via pip
* install pybind11 via pip
* install pymoskito with `python setup.py install`
* go to the example folder tanksystem and run `python __main__.py`

TanDro commented:
> @TanDro do you also experience a segfault after running a simulation with at least one cpp-module?

No, I only get sometimes a segfault by closing pymoskito, but only in normal run. By debugging no segfaults appear. Hence, the debugging is complicated. 

> The current install call just copies the libs back into the original folder (I commented that out for the moment to easier clean the folder). How about the creation of a lib folder in the project root (same level as src) the keep the folder clean?

Yes, that is a good idea, see next commit.

TanDro committed:
Add virtual destructors for ctrl and obs super classes

TanDro committed:
Add additionally lib directory installing the compiled libs

cklb commented:
> No, I only get sometimes a segfault by closing pymoskito, but only in normal run. By debugging no segfaults appear. Hence, the debugging is complicated.

Thats what I meant, only that I can reproduce this in normal mode, with pdb and also in gdb.
However, as I do not have debug info available for the current python version, I do not know for sure where the actual error occurs:

```
29/10/2020 09:20:22 - SimulationGui - INFO - Close Event received, shutting down.

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffd45dab00 in ?? ()
(gdb) bt
#0  0x00007fffd45dab00 in ?? ()
#1  0x00007ffff24b5c21 in cleanup_qobject(_sipSimpleWrapper*, void*) () from /home/cklb/.virtualenvs/pymoskito/lib/python3.8/site-packages/PyQt5/QtCore.abi3.so
#2  0x00007ffff27bdee9 in sip_api_visit_wrappers (visitor=0x7ffff24b5c00 <cleanup_qobject(_sipSimpleWrapper*, void*)>, closure=0x555555585040) at siplib.c:13278
#3  0x00007ffff24b596b in cleanup_on_exit(_object*, _object*) () from /home/cklb/.virtualenvs/pymoskito/lib/python3.8/site-packages/PyQt5/QtCore.abi3.so
#4  0x00007ffff7d196a3 in ?? () from /usr/lib/libpython3.8.so.1.0
#5  0x00007ffff7d2eab2 in PyObject_Call () from /usr/lib/libpython3.8.so.1.0
#6  0x00007ffff7de8799 in ?? () from /usr/lib/libpython3.8.so.1.0
#7  0x00007ffff7c7cb7d in ?? () from /usr/lib/libpython3.8.so.1.0
#8  0x00007ffff7dd2465 in Py_RunMain () from /usr/lib/libpython3.8.so.1.0
#9  0x00007ffff7dae7b9 in Py_BytesMain () from /usr/lib/libpython3.8.so.1.0
#10 0x00007ffff7a4c152 in __libc_start_main () from /usr/lib/libc.so.6
#11 0x000055555555504e in _start ()
```

Basically, all objects get deleted once the simulation is finished and should therefore already be gone when the application is closed. I think the the actual wrapped CPP object behind the wrapper (CppBase) is already freed but the wrapper is not. Then and when Qt does its own cleanup and deletes the wrapper itself, the CPP object is gone and hence we get a SIGSEV.

@simbach Thanks for the information, I'm going to add a little section to the docs on how to set the environment up under windows.

TanDro committed:
Make class CppBase standalone

TanDro committed:
Change order of inheritance for correct work

TanDro commented:
@cklb I found the issue, that causes the segfault. CppBase was also derived from QObject init in the same way the SimulationModule base class. Please can you check, if an error occurs at your setup.
An import information is the inheritance order at the users class.

cklb commented:
Nice work. I just repeated the little stress test that tries to surface the bug (~100 runs of test_tanksystem) and I did not crash in any of the runs. For me it seems like your patch solved it :+1:  

I'm still not sure about the final interface design but the open points should resolve once I finish the "tutorial" part for this new feature. When writing a manual, all quirks and "magic" design choices become visible. ;-)

TanDro committed:
Fix incorrect cmake build command,

TanDro committed:
Change cmake minimum version to 3.14 to avoid vs2017 bug

TanDro commented:
The current version is successful tested on different windows versions (8, 10). It only runs with the usage of build and lib folder as of cmake version 3.14.

cklb commented:
Since the scipy wheels are not yet available for 3.9, pip tries a build from source. This fails since the scipy build dependencies are not installed in the VM. 
As it is not essential for this PR, I would just skip that build for now.

TanDro commented:
Also successful tested on Mac OSX 10.9.5.

cklb commented:
Thats really good news. Concerning the CMake version, in the release notes, it is mentioned that a bump to 3.4 was performed, although pybind11s own `CMake.install` reads:
```
cmake_minimum_required(VERSION 3.4)

# The `cmake_minimum_required(VERSION 3.4...3.18)` syntax does not work with
# some versions of VS that have a patched CMake 3.11. This forces us to emulate
# the behavior using the following workaround:
if(${CMAKE_VERSION} VERSION_LESS 3.18)
  cmake_policy(VERSION ${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION})
else()
  cmake_policy(VERSION 3.18)
endif()
```
I think this is where you got the motivation for the downgrade from?
Would it make sense if we just copy those lines?

TanDro commented:
I set the cmake version to 3.14, because the current Visual Studio Community version (2019) comes with 

`C:\Users\Jens Wurm>cmake --version`
`cmake version 3.14.19060802-MSVC_2`
`CMake suite maintained and supported by Kitware (kitware.com/cmake).`

which runs without problems. Older VS Community versions (<=2017) comes with cmake-3.7, what not work properly.

cklb committed:
Added cmake dep to see which versions are supported on travis

cklb commented:
Alright, so we just required VS 2019 and newer for Windows, I'm fine with that. However, concerning Unix dependencies I would like to stick with Debian stable (as with all other dependencies of pymoskito) and the current version of CMake there is 3.13.4-1.
Furthermore, travis currently uses an ubuntu xenial docker image with cmake version 3.12.4.

I know these are details but would 3.12 be enough?

TanDro commented:
In my opinion linux and version 3.12 shouldn't make any problems, but I can test it today in the evening. 

EDIT: I have testet 3.12 on linux and on windows. On both systems it runs without problems. Hence, 3.12 would be enough.

TanDro committed:
Change cmake minimum version to 3.12

TanDro committed:
Correct values for pid controller

TanDro committed:
Clean up code, add some docu, correct namings, add state controller

TanDro committed:
Summarize of controller and observer implementations in one file and inluding in one python module

TanDro committed:
Rework state controller, add linearize model calculations

TanDro committed:
Finalize controller design

cklb commented:
Nice work, I see the number of controllers is growing ;-)

FYI: If you need a quick way to convert the Jacobians of your RHS for the linearized Approximation, and want to save tedious conversion from Sympy Matrices to Numpy Arrays,
just lambdify the symbolic matrix expressions and evaluate them at the state of your choice.

TanDro committed:
Improve docu of tanksystem example

TanDro committed:
Add auto generated files of pybind binding to gitignore

cklb commented:
I now you are not going to like it but it would be nice if we stick to `snake_case` for function-, method- and property names and only use `CamelCase` for classes like described in PEP8.

I know PyQt and PyQtgraph don't do it and some methods in the GUI modules are still to be updated but the majority of the code base adheres to the standard.

If you want I can also do the renaming part, as I still got some vim macros lying around from the last time ;-)

TanDro commented:
That's not a problem. To be honest I forgot to use this style, because the example is used in another way, too. I don't want to give you any work, what my fault is. But if can share your macros, this would be great ;>

TanDro committed:
Move function/propoerties names to snake_case

TanDro committed:
Add option to add addtional libraries

TanDro committed:
Change handling of last simulations to a normal list

cklb committed:
Ci with actions (#80)

Migrate the CI pipeline from travis to github actions

cklb commented:
Hey, as I'm not allowed to push to this branch, you can find minor additions over at https://github.com/cklb/pymoskito/tree/wip_pybind .
After processing those commits could you please rebase onto develop? I just moved the CI pipeline over to Github actions which should allow Ci tests for this branch.

TanDro committed:
Update os detection to determine mingw

TanDro committed:
Update mingw handling

TanDro committed:
Add comment to use correct pybind version by using additional_lib parameter of CppBase

cklb commented:
I fixed the remaining CI problems. The PR is now (3305c4eac4a5f22d5ecd16bdcfea561cf9577e38) merged in the develop branch. :smile:
